### PR TITLE
Laboratory work #1, Isakov Danila - 21FPL2

### PR DESCRIPTION
**Pull request by vchemsmisl**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -33,8 +47,19 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
 
     In case of corrupt input arguments, None is returned
     """
-    pass
-
+    if isinstance(tokens, list) and isinstance(stop_words, list):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:50]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 2]
```diff
@@ -48,8 +73,17 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
-
+    if isinstance(tokens, list):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:76]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 3]
```diff
@@ -66,8 +100,22 @@ def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[l
 
     In case of corrupt input arguments, None is returned
     """
-    pass
-
+    if isinstance(frequencies, dict) and isinstance(top, int) and not isinstance(top, bool):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:103]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 4]
```diff
@@ -82,7 +130,16 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if frequencies and isinstance(frequencies, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:133]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 5]
```diff
@@ -99,7 +156,21 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if term_freq and isinstance(term_freq, dict) and isinstance(idf, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:159]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 6]
```diff
@@ -118,7 +189,26 @@ def calculate_expected_frequency(
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if doc_freqs and isinstance(doc_freqs, dict) and isinstance(corpus_freqs, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:192]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 7]
```diff
@@ -137,7 +227,19 @@ def calculate_chi_values(expected: dict[str, float], observed: dict[str, int]) -
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if expected and isinstance(expected, dict) and observed and isinstance(observed, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:230]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 8]
```diff
@@ -156,4 +258,15 @@ def extract_significant_words(chi_values: dict[str, float], alpha: float) -> Opt
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if chi_values and isinstance(chi_values, dict) and isinstance(alpha, float):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:261]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 9]
```diff
@@ -17,8 +18,21 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
...
     In case of corrupt input arguments, None is returned
     """
-    pass
-
+    if isinstance(text, str):
+        clean_text_list = []
+        for element in text:
+            if element not in '.,?!;:-()—"\'%#@*&$+=[]{}<>^~`|/':
+                clean_text_list.append(element)
+        if clean_text_list[0] == '':
```
[the remark this answers is not in the document]
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:26]: Done

File: lab_1_keywords_tfidf/main.py [hunk 10]
```diff
@@ -17,8 +18,21 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
-
+    if isinstance(text, str):
```
[the remark this answers is not in the document]
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:21]: Done

File: lab_1_keywords_tfidf/main.py (hunk 1 above)
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:50]: Done

File: lab_1_keywords_tfidf/main.py (hunk 2 above)
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:76]: Done

File: lab_1_keywords_tfidf/main.py (hunk 4 above)
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:133]: Done

File: lab_1_keywords_tfidf/main.py (hunk 5 above)
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:159]: Done

File: lab_1_keywords_tfidf/main.py (hunk 6 above)
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:192]: Done

File: lab_1_keywords_tfidf/main.py (hunk 7 above)
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:230]: Done

File: lab_1_keywords_tfidf/main.py (hunk 8 above)
Comment (vchemsmisl) [reply lab_1_keywords_tfidf/main.py:261]: Done

File: lab_1_keywords_tfidf/main.py [hunk 11]
```diff
@@ -82,7 +129,16 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
...
-    pass
+    if not (frequencies and isinstance(frequencies, dict)):
+        return None
+    for i, j in frequencies.items():
+        if not isinstance(i, str) or not isinstance(j, int):
+            return None
+    word_sum = sum(list(frequencies.values()))
+    words_tf = {}
+    for word, frequency in frequencies.items():
+        words_tf[word] = frequency/word_sum
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:140]:
```suggestion
        words_tf[word] = frequency / word_sum
```

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**